testing/cli/test_cli_sct_dmri_moco.py

```diff
@@ -16,8 +16,8 @@ def test_sct_dmri_moco_check_params(tmp_path):
                              '-ofolder', str(tmp_path)])
 
     lresults = genfromtxt(tmp_path / "moco_params.tsv", skip_header=1, delimiter='\t')[:, 0]
-    lgroundtruth = [0.00047529041677414337, -1.1970542445283172e-05, -1.1970542445283172e-05, -1.1970542445283172e-05,
-                    -0.1296642741802682, -0.1296642741802682, -0.1296642741802682]
+    lgroundtruth = [4.75290417e-04, -1.19489128e-05, -1.19489128e-05, -1.19489128e-05,
+                    -1.29636934e-01, -1.29636934e-01, -1.29636934e-01]
```

Comment:
> So, I think instead, it's much safer to treat the array as the source of truth, and update the header on save/load to reflect the datatype of the array.

Upon making the changes proposed in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3944#discussion_r1021695531, we now only have 2 failing tests:

```
_______________________________________________________________ test_sct_dmri_moco_check_params ________________________________________________________________
Traceback (most recent call last):
  File "/home/joshua/repos/spinalcordtoolbox/testing/cli/test_cli_sct_dmri_moco.py", line 21, in test_sct_dmri_moco_check_params
    assert allclose(lresults, lgroundtruth)
AssertionError: assert False
 +  where False = allclose(array([ 4.75290417e-04, -1.19489128e-05, -1.19489128e-05, -1.19489128e-05,\n       -1.29636934e-01, -1.29636934e-01, -1.29636934e-01]), 
                                 [0.00047529041677414337, -1.1970542445283172e-05, -1.1970542445283172e-05, -1.1970542445283172e-05, -0.1296642741802682, -0.1296642741802682, ...])
__________________________________________________________ test_sct_dmri_moco_with_mask_check_params ___________________________________________________________
Traceback (most recent call last):
  File "/home/joshua/repos/spinalcordtoolbox/testing/cli/test_cli_sct_dmri_moco.py", line 44, in test_sct_dmri_moco_with_mask_check_params
    assert allclose(lresults, lgroundtruth)
AssertionError: assert False
 +  where False = allclose(array([ 0.00804089,  0.00395558,  0.00395558,  0.00395558, -0.01534442,\n       -0.01534442, -0.01534442]), 
                                 [0.008032332623754357, 0.0037734940916436697, 0.0037734940916436697, 0.0037734940916436697, -0.01502861167728611, -0.01502861167728611, ...])
```

The differences are (relatively) small in magnitude. In addition, I've had [issues with these 2 moco tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3075#issuecomment-871366336) in the past:

> If this check fails, it doesn't give us much qualitative feedback:
> 
> * The test fails in an identical way for changes of any magnitude (e.g. 1e-4 and 1e2), so it's hard to gauge the severity of a failing test.
> * The test displays the moco params, but doesn't provide any message describing the values, so it's not immediately obvious what it means for these values to change.
>  * Checking against these values only tests if we're preserving the current state of the moco algorithm. But, I think it would be valuable to also try to capture the goal of motion correction, too. (e.g. Say we started with misaligned data, and had ground truth for aligned data. Then we could check if the moco algorithm performs the realignment to an acceptable degree.)

These ground truth values have been manually updated in the past (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2642). In fact, the ground truth values have even been updated as a direct result of the same issue (#3232) that this PR is trying to fix (see https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3270#discussion_r593780155). 

So, I'm not overly surprised that these tests are failing once more.

Reply:
tl;dr: I think the failing test is actually due to the current "ground truth" values being wrong (due to corruption from scaling issues). And, I think the new values are actually the correct ones!

<details> 
<summary> Investigation </summary>

I set breakpoints within both of the new "datatype mismatch" checks. 

Immediately, I found `float/int` mismatches on these lines

  * https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L244
  
    https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L283
  
    (The `.mean` call causes the `int16`-typed `im` array to change to `float64`)
  
  * https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L612-L614

    (The mathematical operation causes the `int16`-typed `data_targetz` array to change to `float32`.)
  
  * https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/d89f8a3de5253cdca5c5041a467422aa8214fa24/spinalcordtoolbox/moco.py#L377-L378
  
    (Originally, `im_moco` has a `float64` header. However, `im_data` has an `int16` header which gets copied over.)

In all four cases, having a `float` array and an `int` header is what what triggers: 

* #3232. 
 
So, it is likely that the values were being slightly corrupted, and that our supposedly-correct ground truth values... aren't correct? 

After this PR, though, the images are saved with `float64` in both the array _and_ header, which avoids any scaling issues.

</details>

Still, I'd like to dig just a little bit further to see _exactly_ where the changes are introduced. (Are all 4 datatype mismatches contribute? Does one instance have a more outsized effect than the other?)